src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -242,6 +297,140 @@ private static string ValueToHexString(object value)
+            // we can just return "0".
+            if (resultValue == 0)
+            {
+                if (values.Length > 0 && values[0] == 0)
+                {
+                    string enumName = names[0];
+                    if (destination.Length < enumName.Length)
+                    {
+                        charsWritten = 0;
+                        isDestinationTooSmall = true;
+                        return false;
+                    }
+
+                    enumName.CopyTo(destination);
+                    charsWritten = enumName.Length;
+                    isDestinationTooSmall = false;
+                    return true;
+                }
+
+                destination[0] = '0';
+                charsWritten = 1;
+                isDestinationTooSmall = false;
+                return false;
+            }
+
+            // With a ulong result value, regardless of the enum's base type, the maximum
+            // possible number of consistent name/values we could have is 64, since every
+            // value is made up of one or more bits, and when we see values and incorporate
+            // their names, we effectively switch off those bits.
+            Span<int> foundItems = stackalloc int[64];
+
+            // Walk from largest to smallest. It's common to have a flags enum with a single
+            // value that matches a single entry, in which case we can just return the existing
+            // name string.
+            int index = values.Length - 1;
+            while (index >= 0)
+            {
+                if (values[index] == resultValue)
+                {
+                    string enumName = names[index];
+                    if (destination.Length < enumName.Length)
+                    {
+                        charsWritten = 0;
+                        isDestinationTooSmall = true;
+                        return false;
+                    }
+
+                    enumName.CopyTo(destination);
+                    charsWritten = enumName.Length;
+                    isDestinationTooSmall = false;
+                    return true;
+                }
+
+                if (values[index] < resultValue)
+                {
+                    break;
+                }
+
+                index--;
+            }
+
+            // Now look for multiple matches, storing the indices of the values
+            // into our span.
+            int resultLength = 0, foundItemsCount = 0;
+            while (index >= 0)
+            {
+                ulong currentValue = values[index];
+                if (index == 0 && currentValue == 0)
+                {
+                    break;
+                }
+
+                if ((resultValue & currentValue) == currentValue)
+                {
+                    resultValue -= currentValue;
+                    foundItems[foundItemsCount++] = index;
+                    resultLength = checked(resultLength + names[index].Length);
```

Comment:
This scenario is quite exceptional, I think it could be useful for the user to handle this. If they trigger this, it is because they are doing something uncommon. If we just return false, it could be quite hard to diagnose.

I'm not entirely sure about the conventions here regarding exceptions in `TryFormat` methods. However, `decimal.TryParse` can throw a `System.FormatException` for example:
```C#
Span<char> aSpan = stackalloc char[50];
decimal aDecimal = 1;
string anInvalidFormat = "a";
aDecimal.TryFormat(aSpan, out int b, anInvalidFormat);
```